Allow Metrics.NET to export directly to Prometheus.io

clavoie-acquisio opened the pull request:
Since Prometheus has been blessed as "the" way to monitor by the CNCF, it's likely worth adding to Metrics.NET

This is clearly just a WIP, but I was wondering if that's the approach you'd want for this.

PaulParau commented:
I agree, it would be nice to have Prometheus integration with Metrics.NET.

In my opinion the ideal way to add new reporters would be as separate entities from the core metrics library (separate codebase & nuget). This way they can have their own dependencies and lifecycles, and people that don't want to use a certain reporter won't get it when they install the core library.

For Prometheus, I see that you need to have an endpoint that is periodically pulled. Currently there isn't any way of adding a new endpoint directly to the MetricsHttpListener without directly modifying the class. So I'm thinking that in order to implement a decoupled Prometheus reporter we'd have to find a way to add pull-model reporters via config extension methods.

We'll investigate how this can be done (it would also be future-proofing the integration with other pull-model services). Let us know what you think.

clavoie-acquisio commented:
I'm all for decoupling the two, in the meantime, since this is basically
"just another format" I'd dearly appreciate not to have to maintain my own
branch until the future-proofing is ready.

Happy to rebuild this once the plugin reporters are available though.

Thoughts?

On Tue, May 31, 2016 at 10:21 AM, Paul Parau <notifications@github.com>
wrote:

> I agree, it would be nice to have Prometheus integration with Metrics.NET.
>
> In my opinion the ideal way to add new reporters would be as separate
> entities from the core metrics library (separate codebase & nuget). This
> way they can have their own dependencies and lifecycles, and people that
> don't want to use a certain reporter won't get it when they install the
> core library.
>
> For Prometheus, I see that you need to have an endpoint that is
> periodically pulled. Currently there isn't any way of adding a new endpoint
> directly to the MetricsHttpListener without directly modifying the class.
> So I'm thinking that in order to implement a decoupled Prometheus reporter
> we'd have to find a way to add pull-model reporters via config extension
> methods.
>
> We'll investigate how this can be done (it would also be future-proofing
> the integration with other pull-model services). Let us know what you think.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Recognos/Metrics.NET/pull/27#issuecomment-222703646>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AStJnk_grnVYklxPqjwcgoKXn_pZ5KYFks5qHEPvgaJpZM4Io0G3>
> .
>

PaulParau commented:
OK, I understand your point. 

Let's do it this way: since we didn't get a chance to look at the PR in detail we will do that first. If there's an easy/quick way of decoupling the reporter we'll do it, otherwise we'll merge the current approach.

clavoie-acquisio commented:
Sounds good. There's a few more things I want to add to that PR, I just
wanted to surface those issues before I spent too much time on a path you
guys would reject.

In particular, tags should convert to prometheus labels, types need to
convert properly (which is to say, mostly strip them because they have
subtly different semantics :( ) and naming needs to be cleaned up a bit.

Lemme know about decoupling the reporter, and I'll fix the rest before the
weekend. That works for you?

On Tue, May 31, 2016 at 6:47 PM, Paul Parau <notifications@github.com>
wrote:

> OK, I understand your point.
>
> Let's do it this way: since we didn't get a chance to look at the PR in
> detail we will do that first. If there's an easy/quick way of decoupling
> the reporter we'll do it, otherwise we'll merge the current approach.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Recognos/Metrics.NET/pull/27#issuecomment-222844273>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AStJnnTy2OEMZzYj-5IJAYe-K_Kiui8mks5qHLqCgaJpZM4Io0G3>
> .
>

Paccc commented:
I was looking through your code out of curiosity, and I noticed a couple minor things I just wanted to point out. Firstly, there's a typo in the `PromotheusReport.cs` filename. :)  

Secondly, in the reporter, the current timestamp is provided by the `BaseReport`, so you don't need to (and probably shouldn't?) get it using `DateTime.UtcNow`. The values are `base.ReportTimestamp` and `base.CurrentContextTimestamp` which contain the timestamp that the report was started and the timestamp that the `MetricsData` was captured respectively. If you want to use one of those values, the `ReportTimestamp` is set immediately before `StartReport()` is called, and the `CurrentContextTimestamp` is set immediately before `StartContext()` is called. Judging from your code, it looks like you'll likely want to use `ReportTimestamp`.

Additionally, I am also in favor of adding potential support for decoupling pull-model reporters. I had considered implementing this previously but didn't have a strong enough use-case (and available time) to actually do it. However, I think it can be fairly easily and elegantly done using a few new config extensions without making many code changes to the `MetricsHttpListener`.

clavoie-acquisio commented:
On Wed, Jun 1, 2016 at 12:51 AM, Paccc <notifications@github.com> wrote:

> I was looking through your code out of curiosity, and I noticed a couple
> minor things I just wanted to point out. Firstly, there's a typo in the
> PromotheusReport.cs filename. :)
>
Bah. Fixed.

> Secondly, in the reporter, the current timestamp is provided by the
> BaseReport, so you don't need to (and probably shouldn't?) get it using
> DateTime.UtcNow. The values are base.ReportTimestamp and
> base.CurrentContextTimestamp which contain the timestamp that the report
> was started and the timestamp that the MetricsData was captured
> respectively. If you want to use one of those values, the ReportTimestamp
> is set immediately before StartReport() is called, and the
> CurrentContextTimestamp is set immediately before StartContext() is
> called. Judging from your code, it looks like you'll likely want to use
> ReportTimestamp.
>
Actually, I really wanted CurrentContextTimestamp. Fixed.

> Additionally, I am also in favor of adding potential support for
> decoupling pull-model reporters. I had considered implementing this
> previously but didn't have a strong enough use-case (and available time) to
> actually do it. However, I think it can be fairly easily and elegantly done
> using a few new config extensions without making many code changes to the
> MetricsHttpListener.
>
Happy to do it once the Metrics.NET side is in place.

Peace, Love and Metrics,
Christian Lavoie

>

Paccc commented on Src/Metrics/Reporters/PrometheusReport.cs:
```diff
@@ -0,0 +1,188 @@
+            total += ts.Seconds;
+            total *= 1000; // in millis
+            total += ts.Milliseconds;
+            return base.CurrentContextTimestamp >= epoch ? total : -total; // all of this because ts.TotalMilliseconds is a double. *sigh*
```
Just out of curiosity, why can't you do something like `return ts.Ticks - (ts.Ticks % TimeSpan.TicksPerMillisecond);`? If you're worried about floating point inaccuracies, this way keeps everything as a long without having to cast.
Also, what was wrong with the first way you did it, when you casted `ts.TotalMilliseconds` to a long? Doesn't that result in the same value?

clavoie-acquisio replied:
I am worried about floating point inaccuracies, and I just didn't think
about using Ticks. Long story short: brain fart. Fixed.

And what was wrong with ts.TotalMilliseconds was accuracy loss.

On Wed, Jun 1, 2016 at 6:28 PM, Paccc <notifications@github.com> wrote:

> In Src/Metrics/Reporters/PrometheusReport.cs
> <https://github.com/Recognos/Metrics.NET/pull/27#discussion_r65454777>:
>
> > +            reportText = new StringBuilder();
> > +        }
> > +
> > +        private static readonly DateTime epoch = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
> > +        private long CurrentTimeMillis()
> > +        {
> > +            TimeSpan ts = base.CurrentContextTimestamp - epoch;
> > +            ts = ts.Duration(); // absolute value, for dates before Jan 1st, 1970
> > +            long total = ts.Days * 24 + ts.Hours; // total is in hours
> > +            total *= 60; // in minutes
> > +            total += ts.Minutes;
> > +            total *= 60; // in seconds
> > +            total += ts.Seconds;
> > +            total *= 1000; // in millis
> > +            total += ts.Milliseconds;
> > +            return base.CurrentContextTimestamp >= epoch ? total : -total; // all of this because ts.TotalMilliseconds is a double. *sigh*
>
> Just out of curiosity, why can't you do something like return ts.Ticks -
> (ts.Ticks % TimeSpan.TicksPerMillisecond);? If you're worried about
> floating point inaccuracies, this way keeps everything as a long without
> having to cast.
> Also, what was wrong with the first way you did it, when you casted
> ts.TotalMilliseconds to a long? Doesn't that result in the same value?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Recognos/Metrics.NET/pull/27/files/9e562b1d08fcbee54a9da7648a5fd0a6d4e12e54..f5c70cdc1b21f2ec13a25345cde2993640ebb6a0#r65454777>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AStJnlNOSG-Mj65K0A6wmoetHOisxtmZks5qHgecgaJpZM4Io0G3>
> .
>



-- 
Peace, Love and Unfinished Signatures,
Christian "Camembert" Lavoie

PaulParau commented:
I built an initial implementation of support for pull-model reports on the **dev** branch. It's basically a way to add endpoints to the `MetricsHttpListener` at config time by using the `.WithEndpointReporting` method. Individual endpoints can be configured using `MetricsEndpointReports.WithEndpointReport`.

Please have a look and tell me what you think.

clavoie-acquisio commented:
Looks good at first blush, any reason why you didn't convert the existing
reporters? It<s pretty much the final test to see if it's a good API or not.

On Sat, Jun 4, 2016 at 11:39 AM, Paul Parau <notifications@github.com>
wrote:

> I built an initial implementation of support for pull-model reports on the
> *dev* branch. It's basically a way to add endpoints to the
> MetricsHttpListener at config time by using the .WithEndpointReporting
> method. Individual endpoints can be configured using
> MetricsEndpointReports.WithEndpointReport.
>
> Please have a look and tell me what you think.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Recognos/Metrics.NET/pull/27#issuecomment-223762180>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AStJnrS6cY4tWYeTnr22nDSnntRw9YgBks5qIZwngaJpZM4Io0G3>
> .
>



-- 
Peace, Love and Unfinished Signatures,
Christian "Camembert" Lavoie

PaulParau commented:
> any reason why you didn't convert the existing reporters?

Didn't get a chance :) I'll do that soon. I'm also thinking that these changes (new reporter type + prometheus) should be rolled out as prerelease first, just to make sure everything is ironed out before releasing a stable build.

clavoie-acquisio commented:
Just let me know when the first one is converted, I'll port my stuff.
On Jun 6, 2016 17:34, "Paul Parau" <notifications@github.com> wrote:

> any reason why you didn't convert the existing reporters?
>
> Didn't get a chance :) I'll do that soon. I'm also thinking that these
> changes (new reporter type + prometheus) should be rolled out as prerelease
> first, just to make sure everything is ironed out before releasing a stable
> build.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Recognos/Metrics.NET/pull/27#issuecomment-224095670>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AStJnpyOyFxzgzAc4cgLrnBGCtuPZxTiks5qJJJxgaJpZM4Io0G3>
> .
>

PaulParau commented:
I just finished another iteration, you can have a look on the **dev** branch. Among other changes, configuration is now done at HTTP endpoint level. I've also converted the existing endpoints to the new model.

clavoie-acquisio commented:
How do you expect non-default endpoints to be registered?

Is it worth it to think through registering new endpoints without having
access to the source? For the record, it's not a concern I have, but
figured I'd throw it out there.

Otherwise, LGTM.

On Thu, Jun 9, 2016 at 1:50 PM, Paul Parau <notifications@github.com> wrote:

> I just finished another iteration, you can have a look on the *dev*
> branch. Among other changes, configuration is now done at HTTP endpoint
> level. I've also converted the existing endpoints to the new model.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Recognos/Metrics.NET/pull/27#issuecomment-224973735>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AStJngbLI8qu6NYvbw-8OvS2B8B9Y7tYks5qKFJQgaJpZM4Io0G3>
> .
>



-- 
Peace, Love and Unfinished Signatures,
Christian "Camembert" Lavoie

PaulParau commented:
I think a good way to register new endpoints is to do it similarly to the default ones: create an extension method for `MetricsEndpointReports` (e.g. WithPrometheusReport etc.) where you call the `MetricsEndpointReports.WithEndpointReport` (esentially providing a way to generate a `MetricsEndpointResponse`). Then you can call the extension method in the configuration action parameter for every HTTP listener you want (`WithHttpEndpoint` - you can look at the SamplesConsole project for an example which calls `MetricsEndpointReports.WithEndpointReport` directly).

> Is it worth it to think through registering new endpoints without having access to the source?

I think it's best to keep it simple for the moment :) we'll extend the mechanisms for defining and new registering endpoints when/as the need arises.

PaulParau commented:
I've released v0.4.0-pre with support for endpoint reports. I've also documented it here https://github.com/Recognos/Metrics.NET/wiki/Http-Endpoint-Reports.

What do you think about making your own repository for the Prometheus reporter?  (e.g. like the Google Analytics reporter https://github.com/hinteadan/Metrics.NET.GAReporting)

I know I mentioned this before but new reporters should be separate from the core Metrics.NET (I know that elastic search and influxdb reporters are already in there, we'll try to separate them in the future). As a separate project (both repository and nuget package), a custom reporter would have it's own lifecycle and it would be much easier to maintain issues and releases. Adding reporters to the core just bloats it unnecessarily (just imagine having 10 different reporters inside the core, each with its own dependencies).

Let me know what you think.

vic10us commented:
Would you be able to provide a sample implementation of the EndpointReport?

clavoie-acquisio commented:
With apologies for disappearing, that would make my life easier too.

On Mon, Jul 25, 2016 at 10:17 AM, Keith Andrew Hill <
notifications@github.com> wrote:

> Would you be able to provide a sample implementation of the EndpointReport?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Recognos/Metrics.NET/pull/27#issuecomment-234965903>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AStJnorPlVRg_N1wwqj_Mupq6fIIGTaQks5qZMVrgaJpZM4Io0G3>
> .
>



-- 
Peace, Love and Unfinished Signatures,
Christian "Camembert" Lavoie

vic10us commented:
Looks like doing something like this works.
```
config.WithEndpointReport("/prometheus", (d,h,c) => new MetricsEndpointResponse(PrometheusReport.RenderMetrics(d,h), "text/plain"))
```
Problem is that it only works with HttpEndpoints. The Owin adapter is still using older logic. :(

paulkling commented:
I was curious on this PR. Do we think this will be merged at some point? What will it take to merge it? 

I am super excited to try this feature out.

PaulParau commented:
> Problem is that it only works with HttpEndpoints. The Owin adapter is still using older logic. :(

@vic10us Indeed, the Owin adapter does not support this new model, nor does the Nancy adapter for that matter. It's something I overlooked & I intend to look into it as soon as I can.

The example provided by @vic10us is correct. Support for endpoint reports is still in beta because I wanted to see an actual implementation of such a report (besides the built-in reports which were already ported), in case changes were still needed to the API (they might still be needed - re: adapters). 

In the end, to merge the PR, it should be moved to a separate csproj & made to use the new endpoint reporter model (which was build specifically to allow such separation). @clavoie-acquisio you can try to port it now, you can wait until after I settle the issues with the adapter or if you want I can do it myself when I get some time. Or, if you want to, you could make & maintain a separate repo & nuget package (which would have the advantages I stated in my post above). It is up to you :)

clavoie-acquisio commented:
It's a question of finding the time. I aim to get a new repo "ASAP", but
I've been beyond swamped. First to get to it? ;)

On Fri, Jul 29, 2016 at 11:20 AM, Paul Parau <notifications@github.com>
wrote:

> Problem is that it only works with HttpEndpoints. The Owin adapter is
> still using older logic. :(
>
> @vic10us <https://github.com/vic10us> Indeed, the Owin adapter does not
> support this new model, nor does the Nancy adapter for that matter. It's
> something I overlooked & I intend to look into it as soon as I can.
>
> The example provided by @vic10us <https://github.com/vic10us> is correct.
> Support for endpoint reports is still in beta because I wanted to see an
> actual implementation of such a report (besides the built-in reports which
> were already ported), in case changes were still needed to the API (they
> might still be needed - re: adapters).
>
> In the end, to merge the PR, it should be moved to a separate csproj &
> made to use the new endpoint reporter model (which was build specifically
> to allow such separation). @clavoie-acquisio
> <https://github.com/clavoie-acquisio> you can try to port it now, you can
> wait until after I settle the issues with the adapter or if you want I can
> do it myself when I get some time. Or, if you want to, you could make &
> maintain a separate repo & nuget package (which would have the advantages I
> stated in my post above). It is up to you :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Recognos/Metrics.NET/pull/27#issuecomment-236209442>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AStJnp7nY_zvoSz4fnGQcaW8ukjULgaPks5qahoigaJpZM4Io0G3>
> .
>



-- 
Peace, Love and Unfinished Signatures,
Christian "Camembert" Lavoie

vic10us commented:
I'm ready to do it. Gimme a little time. :)

PaulParau commented:
No rush :) I also don't know when I'll get to fixing the adapter issues, I hope soon. Hopefully there won't be any more major API changes.

vic10us commented:
I've decided to push this to my own repo https://github.com/DEXSinc/Metrics.NET.Prometheus and created another Nuget Package Metrics.NET.Prometheus it's built against the pre-release and has the same numbering scheme. currently 0.4.1-pre

PaulParau commented:
Wow, that was quick. Great job!

clavoie-acquisio commented:
Thanks Keith.

On Sat, Jul 30, 2016 at 6:36 AM, Paul Parau <notifications@github.com>
wrote:

> Wow, that was quick. Great job!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Recognos/Metrics.NET/pull/27#issuecomment-236358184>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AStJnrHi5gcFGg5d60q83diTiN6hW6VBks5qaykUgaJpZM4Io0G3>
> .
>



-- 
Peace, Love and Unfinished Signatures,
Christian "Camembert" Lavoie

PaulParau commented:
I created a page in the wiki (also linked from the readme) with a list of external reporters and added a link to the Prometheus reporter. Since it was implemented as a separate repository, I'm closing this PR.

I've created a new issue #37 for adding adapter support.

Thanks again @clavoie-acquisio & @vic10us !
